Update `naked` to use the attribute template

ehuss opened the pull request:
New rules:
- ❗ `attributes.codegen.naked.syntax`
- ❗ `attributes.codegen.naked.allowed-positions`
- ❗ `attributes.codegen.naked.duplicates`
- ❗ `attributes.codegen.naked.target_feature`
- ❗ `attributes.codegen.naked.abi`

Renamed rules:
- `attributes.codegen.naked.unsafe-attribute` is now `attributes.codegen.naked.unsafe`

ehuss committed:
Add naked example

ehuss committed:
Add naked attribute template rules

ehuss committed:
Move and reword attributes.codegen.naked.unsafe

To follow the attribute template, and to simplify the wording a little.

ehuss committed:
Add attributes.codegen.naked.target_feature

naked cannot be used with target_feature.
https://github.com/rust-lang/rust/blob/aae43c4532690153af7465227816c93036bb1604/compiler/rustc_passes/src/check_attr.rs#L672-L685

ehuss committed:
Add attributes.codegen.naked.abi

Naked attributes cannot have the "Rust" ABI.
https://github.com/rust-lang/rust/blob/aae43c4532690153af7465227816c93036bb1604/compiler/rustc_passes/src/check_attr.rs#L641-L652

tshepang commented on src/attributes/codegen.md:
```diff
@@ -52,15 +52,51 @@ r[attributes.codegen.naked]
+- [Free functions][items.fn]
+- [Inherent associated functions][items.associated.fn]
+- [Trait impl functions][items.impl.trait]
+- [Trait definition functions][items.traits] with a body
```
does this work
```suggestion
- [Trait][items.traits] [associated functions][items.associated.fn] with a body
```

ehuss replied:
This wording can be ambiguous, since it is unclear if it is an associated function in a trait definition or a trait impl. The original wording was specifically chosen to avoid that ambiguity and contrast with the previous item.

tshepang replied:
initially wanted to link this to `items.traits.associated-item-decls`, due to being more specific, because `items.traits` was too general

> - [Trait definition functions][items.traits.associated-item-decls] with a body

is that alright

ehuss committed:
Minor update of `naked`

More closely align with the template, and some minor word tweaks.

ehuss committed:
Use more specific link for trait definition function

traviscross committed:
Revise `naked` attribute text

Let's revise the text in the `naked` attribute section with some
editorial adjustments.
